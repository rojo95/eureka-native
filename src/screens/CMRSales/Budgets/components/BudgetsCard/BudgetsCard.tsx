import {
    StyleSheet,
    TouchableOpacityProps,
    View,
    TouchableOpacity,
} from "react-native";
import Text from "components/Text/Text";
import { ActivityIndicator, DefaultTheme, useTheme } from "react-native-paper";
import CustomBadge from "components/CustomBadge/CustomBadge";
import { formatPrices } from "utils/numbers";
import { useContext, useMemo } from "react";
import { getColorState } from "../../utils/utils";
import { UserContext } from "contexts/UserContext";

interface BudgetsCard extends TouchableOpacityProps {
    number?: string;
    title: string;
    state: { id: number; name: string };
    totalCost: number;
    totalSale: number;
    onPress?: () => void;
}

const BudgetsCard = ({
    number,
    title,
    state,
    totalCost,
    totalSale,
    onPress,
}: BudgetsCard) => {
    const { language } = useContext(UserContext);
    const theme: DefaultTheme = useTheme();

    const themedStyles = StyleSheet.create({
        container: {
            backgroundColor: theme.colors.primaryContrast,
        },
        title: {
            color: theme.colors.dark,
        },
        code: {
            color: theme.colors.codeColor,
        },
        totalCost: {
            color: theme.colors.danger,
        },
        totalSale: {
            color: theme.colors.success,
        },
    });

    const formattedTotalCost = useMemo(() => {
        return formatPrices({ number: totalCost || 0, language });
    }, [totalCost, language]);

    const formattedTotalSale = useMemo(() => {
        return formatPrices({ number: totalSale || 0, language });
    }, [totalSale, language]);

    return (
        <TouchableOpacity
            style={{ flex: 1, alignItems: "center", justifyContent: "center" }}
            onPress={onPress}
        >
            <View style={[styles.container, themedStyles.container]}>
                <View style={[styles.content]}>
                    <View>
                        <Text style={[styles.code, themedStyles.code]}>
                            {number}
                        </Text>
                    </View>
                    <Text style={[styles.bold, themedStyles.title]}>
                        {title}
                    </Text>
                    <CustomBadge
                        customStyles={{
                            backgroundColor: getColorState({
                                statusId: state.id,
                                theme,
                            }),
                            color: theme.colors.dark,
                        }}
                    >
                        {state.name}
                    </CustomBadge>
                    <Text style={[styles.number]}>
                        {formattedTotalCost === "" ? (
                            <ActivityIndicator size="small" />
                        ) : (
                            <Text>
                                <Text
                                    style={[
                                        styles.bold,
                                        themedStyles.totalCost,
                                    ]}
                                >
                                    {formattedTotalCost}€
                                </Text>{" "}
                                -{" "}
                                <Text
                                    style={[
                                        styles.bold,
                                        themedStyles.totalSale,
                                    ]}
                                >
                                    {formattedTotalSale}€
                                </Text>
                            </Text>
                        )}
                    </Text>
                </View>
            </View>
        </TouchableOpacity>
    );
};

export default BudgetsCard;

const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
        borderTopLeftRadius: 0,
        borderBottomStartRadius: 0,
        borderRadius: 10,
        overflow: "hidden",
        shadowColor: "#000",
        elevation: 5,
        marginVertical: 10,
        marginEnd: 10,
        alignSelf: "center",
    },
    content: {
        flex: 1,
        padding: 16,
    },
    title: {
        fontSize: 18,
        marginBottom: 8,
    },
    number: {
        fontSize: 14,
        textAlign: "right",
    },
    code: {
        fontSize: 14,
    },
    bold: {
        fontWeight: "bold",
    },
});
