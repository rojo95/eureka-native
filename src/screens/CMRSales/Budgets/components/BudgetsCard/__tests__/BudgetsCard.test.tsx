import React from "react";
import renderer from "react-test-renderer";
import BudgetsCard from "../BudgetsCard";

describe("Test to evaluate the <BudgetsCard/> component", () => {
    it("should render correctly", () => {
        const tree: any = renderer
            .create(
                <BudgetsCard
                    title={""}
                    state={{ id: 0, name: "" }}
                    totalSale={0}
                    totalCost={0}
                />
            )
            .toJSON();
        expect(tree.children.length).toBe(1);
    });

    it("should render correctly the totalCost value", () => {
        const totalCost = 500;

        const tree: any = renderer
            .create(
                <BudgetsCard
                    title={""}
                    state={{ id: 0, name: "" }}
                    totalSale={0}
                    totalCost={totalCost}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[3]
                ?.children[0].children;
        expect(componentBase).toEqual([`${totalCost}`, "€"]);
    });

    it("should render correctly the sell value", () => {
        const totalSale = 600;

        const tree: any = renderer
            .create(
                <BudgetsCard
                    title={""}
                    state={{ id: 0, name: "" }}
                    totalCost={10}
                    totalSale={totalSale}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[3]
                ?.children[3].children;
        expect(componentBase).toEqual([`${totalSale}`, "€"]);
    });

    it("should render correctly the state text", () => {
        const testStatusText = "progress";

        const tree: any = renderer
            .create(
                <BudgetsCard
                    title={""}
                    totalSale={0}
                    totalCost={0}
                    state={{ id: 1, name: testStatusText }}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[2]
                ?.children[0]?.children[0];
        expect(componentBase).toContain(testStatusText);
    });

    it("should render correctly the index code", () => {
        const idTest = "1234";

        const tree: any = renderer
            .create(
                <BudgetsCard
                    number={idTest}
                    title={""}
                    totalSale={0}
                    totalCost={0}
                    state={{ id: 0, name: "" }}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[0]
                ?.children[0]?.children[0];
        expect(componentBase).toContain(idTest);
    });

    it("should render correctly the title", () => {
        const testDesc = "test title";

        const tree: any = renderer
            .create(
                <BudgetsCard
                    title={testDesc}
                    totalSale={0}
                    totalCost={0}
                    state={{ id: 1, name: "" }}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[1]?.children;
        expect(componentBase).toContain(testDesc);
    });

    it("render card without change", () => {
        const tree = renderer
            .create(
                <BudgetsCard
                    title={""}
                    totalSale={0}
                    totalCost={0}
                    state={{ id: 1, name: "" }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
