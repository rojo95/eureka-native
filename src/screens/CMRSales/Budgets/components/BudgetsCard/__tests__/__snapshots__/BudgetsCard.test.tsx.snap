// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to evaluate the <BudgetsCard/> component render card without change 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "borderBottomStartRadius": 0,
          "borderRadius": 10,
          "borderTopLeftRadius": 0,
          "elevation": 5,
          "flexDirection": "row",
          "marginEnd": 10,
          "marginVertical": 10,
          "overflow": "hidden",
          "shadowColor": "#000",
        },
        {
          "backgroundColor": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
            "padding": 16,
          },
        ]
      }
    >
      <View>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "fontSize": 14,
                },
                {
                  "color": undefined,
                },
              ],
            ]
          }
        />
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            [
              {
                "fontWeight": "bold",
              },
              {
                "color": undefined,
              },
            ],
          ]
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 8,
          }
        }
      >
        <Text
          collapsable={false}
          numberOfLines={1}
          style={
            {
              "alignSelf": "flex-end",
              "backgroundColor": "rgba(179, 38, 30, 1)",
              "borderRadius": 10,
              "color": undefined,
              "fontSize": 10,
              "height": 20,
              "lineHeight": 10,
              "marginRight": 8,
              "minWidth": 20,
              "opacity": 1,
              "overflow": "hidden",
              "paddingHorizontal": 3,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            [
              {
                "fontSize": 14,
                "textAlign": "right",
              },
            ],
          ]
        }
      >
        <View
          accessibilityRole="progressbar"
          accessibilityState={
            {
              "busy": true,
            }
          }
          accessible={true}
          style={
            [
              {
                "alignItems": "center",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "height": 24,
                "opacity": 1,
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "bottom": 0,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "transform": [
                      {
                        "rotate": "45deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "height": 24,
                        "transform": [
                          {
                            "translateY": 0,
                          },
                          {
                            "rotate": "-165deg",
                          },
                        ],
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "height": 12,
                          "overflow": "hidden",
                          "width": 24,
                        }
                      }
                    >
                      <View
                        collapsable={false}
                        style={
                          {
                            "borderColor": "rgba(103, 80, 164, 1)",
                            "borderRadius": 12,
                            "borderWidth": 2.4,
                            "height": 24,
                            "width": 24,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "bottom": 0,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "transform": [
                      {
                        "rotate": "45deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 12,
                      "overflow": "hidden",
                      "top": 12,
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "height": 24,
                        "transform": [
                          {
                            "translateY": -12,
                          },
                          {
                            "rotate": "345deg",
                          },
                        ],
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "height": 12,
                          "overflow": "hidden",
                          "width": 24,
                        }
                      }
                    >
                      <View
                        collapsable={false}
                        style={
                          {
                            "borderColor": "rgba(103, 80, 164, 1)",
                            "borderRadius": 12,
                            "borderWidth": 2.4,
                            "height": 24,
                            "width": 24,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </Text>
    </View>
  </View>
</View>
`;
